syntax = "proto3";

service Bootstrap {
    rpc Initialize (BootstrapInit) returns (BootstrapInitReply);
}

service Game {
    rpc MissileApproaching (GameMissileApproaching) returns (GamePosition);
    rpc PositionStatus (GamePositionStatus) returns (GamePositionStatusReply);
    rpc Status (GameSoldierStatus) returns (GameSoldierWasHit);
    rpc TransferCommander (CommanderParams) returns (Empty);
    rpc GameOver (Empty) returns (Empty);
}

message BootstrapInit {
    string ip = 1;
    int32 port = 2;
}

message BootstrapInitReply {
    int32 board_size = 1;
}

message GamePosition {
    int32 x = 1;
    int32 y = 2;
}

message GameMissileApproaching {
    GamePosition position = 1;
    int32 time_to_approach = 3;
    int32 missile_type = 4;
}

message GamePositionStatus {
    bool is_occupied = 1;
}

message GamePositionReady {}

message GamePositionStatusReply {
    oneof status {
        GamePosition position = 1;
        GamePositionReady ready = 2;
    }
}

message GameSoldierStatus {}

message GameSoldierWasHit {
    bool was_hit = 1;
}

message SoldierParams {
    string s_id = 1;
    string ip = 2;
    int32 port = 3;
    GamePosition pos = 4;
}

message CommanderParams {
    int32 M = 1;
    int32 ct = 2;
    int32 t = 3;
    int32 T = 4;
    GameMissileApproaching m = 5;
    repeated SoldierParams soldiers = 9;
}

message Empty {}